### PR TITLE
[CQT-66] Create examples for ABA decomposer

### PR DESCRIPTION
**Pull request by juanboschero**
- Create documentation for all decomposition strategies 
-  Include  example usage 
- Create corresponding tests

### CODE REVIEW

**Review by S-Linde — CHANGES_REQUESTED** [review]
Before I go into a more detailed review I want to mention some more high level stuff. The tutorial is very detailed and contains a lot of math. This level of explanation is great, but it removes some of the flow of your text. My suggestion would be to keep the mathematical explanation in the notebook, but put it at the end and refer to it in the text.

Then you can explain on a high level what a ABA decomposer does and then dive straight into the examples, which is what I would expect from a tutorial. The interested reader is then referred to the math further below.

File: pyproject.toml [hunk 1]
```diff
@@ -37,6 +37,7 @@ numpy = [
 ]
 libqasm = "0.6.6"
 networkx = "^3.0.0"
+jupyter = "^1.0.0"
```
[the remark this answers is not in the document]
Comment (juanboschero) [reply pyproject.toml:40]: Oh nice catch! It should be with sympy in the optionals deps

**Review by S-Linde — COMMENTED** [review]
Looking very nice! This structure is much more digestible for me. Really like the ascii-art circuits! (Might be nice to have that for the circuits in the examples as well?)

File: example/tutorials/decompositions.ipynb [hunk 2]
```diff
@@ -0,0 +1,860 @@
...
+     "end_time": "2024-07-23T09:06:24.555910Z",
+     "start_time": "2024-07-23T09:06:24.544193Z"
+    }
+   },
+   "cell_type": "code",
+   "source": [
+    "from opensquirrel.decomposer.aba_decomposer import ZYZDecomposer\n",
+    "\n",
+    "#Decompose the circuit using ZYZDecomposer           \n",
+    "circuit.decompose(decomposer = ZYZDecomposer())\n",
```
[the remark this answers is not in the document]
Comment (juanboschero) [reply example/tutorials/decompositions.ipynb:96]: Do you know how to add this in the `pyproject.toml`?

File: test/example/tutorials/test_decompositions.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+from opensquirrel.circuit_builder import CircuitBuilder
+from opensquirrel.decomposer.aba_decomposer import XZXDecomposer, ZYZDecomposer
+from opensquirrel.decomposer.cnot_decomposer import CNOTDecomposer
+from opensquirrel.decomposer.mckay_decomposer import McKayDecomposer
+from opensquirrel.default_gates import H, Rx, Rz
+from opensquirrel.ir import Float, Qubit
+
+
+class TestDecomposition:
+    def test_sympy_disp(self):
```
Comment (rturrado) [inline test/example/tutorials/test_decompositions.py:15]: What is `disp`? I would change the name into  `test_simpy_<whatever full word>`.

File: test/example/tutorials/test_decompositions.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+        y1 = sp.algebras.Quaternion.from_axis_angle((0, 1, 0), theta1)
+        z = sp.algebras.Quaternion.from_axis_angle((0, 0, 1), theta2)
+        y2 = sp.algebras.Quaternion.from_axis_angle((0, 1, 0), theta3)
+
+        rhs = sp.trigsimp(sp.expand(y1 * z * y2))
+        rhs_string = "cos(theta_2/2)*cos((theta_1 + theta_3)/2) + sin(theta_2/2)*sin((theta_1 - theta_3)/2)*i + sin((theta_1 + theta_3)/2)*cos(theta_2/2)*j + sin(theta_2/2)*cos((theta_1 - theta_3)/2)*k"
+        assert str(rhs) == rhs_string
+
+    def test_builder_ABA(self):
```
Comment (rturrado) [inline test/example/tutorials/test_decompositions.py:54]: What does `builder_ABA` means here? What is being tested? The builder? An input circuit for the ABA decomposer?

I would suggest to not use uppercase for function names.

File: test/example/tutorials/test_decompositions.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+Ry(1.0471976) q[0]
+Rz(-1.5707963) q[0]
+"""
+        )
+        assert (
+            XZXDecomposer().decompose(H(Qubit(0)))
+            == [Rx(Qubit(0), Float(math.pi/2)), Rz(Qubit(0), Float(math.pi/2)), Rx(Qubit(0), Float(math.pi/2))]
+        )
+
+    def test_builder_and_mckay(self):
```
Comment (rturrado) [inline test/example/tutorials/test_decompositions.py:123]: Test this and that sounds a bit like a code smell. Maybe just `tests_mckay_decomposer`?

File: test/example/tutorials/test_decompositions.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+X90 q[0]
+Rz(-1.5707963) q[0]
+X90 q[0]
+Rz(2.0943951) q[0]
+X90 q[0]
+Rz(-1.5707963) q[0]
+"""
+        )
+
+    def test_builder_and_cnot(self):
```
Comment (rturrado) [inline test/example/tutorials/test_decompositions.py:173]: Same here: `test_cnot_decomposer`?

**Review by rturrado — COMMENTED** [review]
They say naming is one of the most difficult things when programming.

But decent, consistent names gives us a hint that there is some structure in the code.

Non descriptive, inconsistent names could indicate the opposite. Here, that's a false negative, because the tests are actually well structured.

So I'm just missing some better names.

**Comment by rturrado** [comment]
It seems:
1. The branch name has nothing to do with the task name.
2. The goal Jira task, to which name I've just set the name of this PR, has been updated from 'Add examples for ABA decomposer' to 'Add examples for all the decompositions'.

File: test/example/tutorials/test_decompositions.py (hunk 4 above)
Comment (rturrado) [reply test/example/tutorials/test_decompositions.py:54]: According to the names of the other tests, this should be `test_aba_decomposer`:

```suggestion
    def test_aba_decomposer(self):
```

File: test/example/tutorials/test_decompositions.py (hunk 5 above)
Comment (rturrado) [reply test/example/tutorials/test_decompositions.py:123]:
```suggestion
    def test_mckay_decomposer(self):
```

File: test/example/tutorials/test_decompositions.py (hunk 3 above)
Comment (rturrado) [reply test/example/tutorials/test_decompositions.py:15]:
```suggestion
    def test_sympy_output(self):
```

Comment (rturrado) [reply test/example/tutorials/test_decompositions.py:15]: Why would we test sympy code?

File: test/example/tutorials/test_decompositions.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+        y1 = sp.algebras.Quaternion.from_axis_angle((0, 1, 0), theta1)
+        z = sp.algebras.Quaternion.from_axis_angle((0, 0, 1), theta2)
+        y2 = sp.algebras.Quaternion.from_axis_angle((0, 1, 0), theta3)
+
+        rhs = sp.trigsimp(sp.expand(y1 * z * y2))
+        rhs_string = "cos(theta_2/2)*cos((theta_1 + theta_3)/2) + sin(theta_2/2)*sin((theta_1 - theta_3)/2)*i + sin((theta_1 + theta_3)/2)*cos(theta_2/2)*j + sin(theta_2/2)*cos((theta_1 - theta_3)/2)*k"
+        assert str(rhs) == rhs_string
+
+    def test_aba_decomposer_builder_to_circuit(self):
```
Comment (rturrado) [inline test/example/tutorials/test_decompositions.py:54]:
```suggestion
    def test_aba_decomposer_input_circuit_creation(self):
    """This test uses the `CircuitBuilder` to build a circuit that serves as input to the ABA decomposer."""
```

File: test/example/tutorials/test_decompositions.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+        theta1, theta2, theta3 = sp.symbols("theta_1 theta_2 theta_3")
+
+        y1 = sp.algebras.Quaternion.from_axis_angle((0, 1, 0), theta1)
+        z = sp.algebras.Quaternion.from_axis_angle((0, 0, 1), theta2)
+        y2 = sp.algebras.Quaternion.from_axis_angle((0, 1, 0), theta3)
+
+        rhs = sp.trigsimp(sp.expand(y1 * z * y2))
+        rhs_string = "cos(theta_2/2)*cos((theta_1 + theta_3)/2) + sin(theta_2/2)*sin((theta_1 - theta_3)/2)*i + sin((theta_1 + theta_3)/2)*cos(theta_2/2)*j + sin(theta_2/2)*cos((theta_1 - theta_3)/2)*k"
+        assert str(rhs) == rhs_string
```
[the remark this answers is not in the document]
Comment (juanboschero) [reply test/example/tutorials/test_decompositions.py:52]: It is part of the Jupyter example, if Sympy updates or becomes incompatible with newer dependencies then we will have to know.

[the remark this answers is not in the document]
Comment (rturrado) [reply test/example/tutorials/test_decompositions.py:52]: Fair enough!

**Review by S-Linde — APPROVED** [review]

**Review by rturrado — APPROVED** [review]

**Review by S-Linde — APPROVED** [review]

File: example/README.md [hunk 9]
```diff
@@ -1,3 +1,27 @@
-Code and tests are the best documentation.
+The documentation for the code can be found in various Jupyter notebooks located in the `tutorials` folder.
 
-For example usage, see `test/test_integration.py`.
\ No newline at end of file
+`Jupyter` and `sympy` are required to access the notebooks. These can be installed all at once by running.
```
Comment (rturrado) [inline example/README.md:3]:
```suggestion
`Jupyter` and `sympy` are required to access the notebooks. These can be installed all at once by running:
```

File: example/README.md [hunk 10]
```diff
@@ -1,3 +1,27 @@
...
+pip install jupyter
+```
+
+and `sympy`,
+
+```
+pip install sympy
+```
+
+To open the `Jupyter` notebooks, one can run,
```
Comment (rturrado) [inline example/README.md:21]:
```suggestion
To open the `Jupyter` notebooks, one can run
```

File: example/README.md [hunk 11]
```diff
@@ -1,3 +1,27 @@
...
+
+```
+pip install jupyter
+```
+
+and `sympy`,
+
+```
+pip install sympy
+```
```
Comment (rturrado) [inline example/README.md:19]:
```suggestion
Alternatively, one can install both packages manually:

```
pip install jupyter sympy
```
```

File: example/README.md [hunk 12]
```diff
@@ -1,3 +1,27 @@
...
 
-For example usage, see `test/test_integration.py`.
\ No newline at end of file
+`Jupyter` and `sympy` are required to access the notebooks. These can be installed all at once by running.
+
+```
+pip install opensquirrel[examples]
+```
+
+Alternatively, one can install both packages manually. Installing `Jupyter` using pip can be done by executing,
```
Comment (rturrado) [inline example/README.md:9]:
```suggestion
Alternatively, one can install both packages manually:
```

File: example/README.md [hunk 13]
```diff
@@ -1,3 +1,27 @@
...
+
+```
+pip install jupyter
```
Comment (rturrado) [inline example/README.md:12]:
```suggestion
pip install jupyter sympy
```

File: example/README.md (hunk 11 above)
Comment (rturrado) [inline example/README.md:19]:
```suggestion
```

**Review by elenbaasc — APPROVED** [review]

**Pull request merged**